packages/next/src/client/components/router-reducer/reducers/server-action-reducer.ts

```diff
@@ -131,113 +135,125 @@ export function serverActionReducer(
-      if (actionFlightData) {
-        const href = createHrefFromUrl(redirectLocation, false)
-        const previousCacheEntry = state.prefetchCache.get(href)
-        state.prefetchCache.set(href, {
-          data: createRecordFromThenable(
-            Promise.resolve([
-              actionFlightData,
-              // TODO-APP: verify the logic around canonical URL overrides
-              undefined,
-            ])
-          ),
-          kind: previousCacheEntry?.kind ?? PrefetchKind.TEMPORARY,
-          prefetchTime: Date.now(),
-          treeAtTimeOfPrefetch: action.mutable.previousTree!,
-          lastUsedTime: null,
-        })
-      }
-
-      // we throw the redirection in the action handler so that it is caught during render
-      action.reject(
-        getRedirectError(redirectLocation.toString(), RedirectType.push)
+      return handleExternalUrl(
+        state,
+        mutable,
+        redirectLocation.href,
+        state.pushRef.pendingPush
       )
-    } else {
-      if (actionFlightData) {
-        const href = createHrefFromUrl(
-          new URL(action.mutable.previousUrl!, window.location.origin),
-          false
-        )
-        const previousCacheEntry = state.prefetchCache.get(href)
-        state.prefetchCache.set(
-          createHrefFromUrl(
-            new URL(action.mutable.previousUrl!, window.location.origin),
-            false
-          ),
-          {
-            data: createRecordFromThenable(
-              Promise.resolve([
-                actionFlightData,
-                // TODO-APP: verify the logic around canonical URL overrides
-                undefined,
-              ])
-            ),
-            kind: previousCacheEntry?.kind ?? PrefetchKind.TEMPORARY,
-            prefetchTime: Date.now(),
-            treeAtTimeOfPrefetch: action.mutable.previousTree!,
-            lastUsedTime: null,
-          }
-        )
-        // this is an intentional hack around React: we want to update the tree in a new render
-        setTimeout(() => {
-          action.changeByServerResponse(
-            action.mutable.previousTree!,
-            actionFlightData,
-            // TODO-APP: verify the logic around canonical URL overrides
-            undefined
-          )
-        })
-      }
+    }
+    return state
+  }
+
+  if (typeof flightData === 'string') {
+    // Handle case when navigating to page in `pages` from `app`
+    return handleExternalUrl(
+      state,
+      mutable,
+      flightData,
+      state.pushRef.pendingPush
+    )
+  }
+
+  // Remove cache.data as it has been resolved at this point.
+  mutable.inFlightServerAction = null
+
+  for (const flightDataPath of flightData) {
+    // FlightDataPath with more than two items means unexpected Flight data was returned
+    if (flightDataPath.length !== 3) {
+      // TODO-APP: handle this case better
+      console.log('SERVER ACTION APPLY FAILED')
```

Comment:
we should throw an invariant here

Reply:
This case is intentionally not throwing similarly to `router.refresh()`. It's something we'll probably want to error on in the future but even then with middleware you could rewrite a path and cause this case to happen.